rust/scx_utils/src/load_balance.rs

```diff
@@ -0,0 +1,318 @@
+//! Assume we're on a 16-core (32-CPU) host with two core complexes:
+//!
+//!```
+//!     let mut aggregator = LoadBalancer::aggregator(32, false);
+//!     aggregator.record_dom_load(0, 1, 1.0);
+//!     aggregator.record_dom_load(1, 1, 1.0);
+//!     // ...
+//!     aggregator.record_dom_load(31, 1, 1.0);
+//!     // ...
+//!     aggregator.record_dom_load(63, 1, 1.0);
+//!     aggregator.record_dom_load(64, 10000, 1.0);
+//!
+//!     let balancer = aggregator.calculate();
+//!
+//!     // Outputs: 66.06451612903226
+//!     info!("{}", balancer.global_load_sum());
+//!```
+//!
+//! For more details on the infeasible weights problem, please see the following
+//! document:
+//!
+//! https://drive.google.com/file/d/1fAoWUlmW-HTp6akuATVpMxpUpvWcGSAv
+
+use anyhow::bail;
+use anyhow::Result;
+use std::collections::BTreeMap;
+
+const MIN_WEIGHT: usize = 1;
+const MAX_WEIGHT: usize = 10000;
+
+#[derive(Debug)]
+pub struct LoadBalancer {
+    dom_load_sums: Vec<f64>,
+    dom_dcycle_sums: Vec<f64>,
+    global_dcycle_sum: f64,
+    global_load_sum: f64,
+    w_prime: f64,
+}
+
+impl LoadBalancer {
+    /// Create a LoadAggregator object that can be given domains and loads by
+    /// the caller, and then used to create a LoadBalancer object.
+    pub fn aggregator(nr_cpus: usize, dcycle_only: bool) -> LoadAggregator {
+        LoadAggregator {
+            doms: BTreeMap::new(),
+            global_loads: BTreeMap::new(),
+            nr_cpus,
+            max_weight: 0,
+            global_dcycle_sum: 0.0f64,
+            global_load_sum: 0.0f64,
+            w_prime: 10000.0f64,
+            dcycle_only,
+        }
+    }
+
+    /// Return the global, host-wide sum of duty cycles.
+    pub fn global_dcycle_sum(&self) -> f64 {
+        self.global_dcycle_sum
+    }
+
+    /// Return the global, host-wide load sum; adjusted for infeasibility.
+    pub fn global_load_sum(&self) -> f64 {
+        self.global_load_sum
+    }
+
+    /// Return an array of domain duty cycle sums, indexed by ID.
+    pub fn dom_dcycle_sums(&self) -> &[f64] {
+        &self.dom_dcycle_sums
+    }
+
+    /// Return an array of domain load sums, indexed by ID, and adjusted for
+    /// infeasibility.
+    pub fn dom_load_sums(&self) -> &[f64] {
+        &self.dom_load_sums
+    }
+
+    /// If applicable, return the adjusted weight for all infeasible scheduling
+    /// entities.
+    pub fn w_prime(&self) -> f64 {
+        self.w_prime
+    }
+}
+
+#[derive(Debug)]
+struct Domain {
+    loads: BTreeMap<usize, f64>,
+    dcycle_sum: f64,
+    load_sum: f64,
+}
+
+fn approx_eq(a: f64, b: f64) -> bool {
+    (a - b).abs() <= 0.0001f64
+}
+
+fn approx_ge(a: f64, b: f64) -> bool {
+    a > b || approx_eq(a, b)
+}
+
+#[derive(Debug)]
+pub struct LoadAggregator {
+    doms: BTreeMap<usize, Domain>,
+    global_loads: BTreeMap<usize, f64>,
+    nr_cpus: usize,
+    max_weight: usize,
+    global_dcycle_sum: f64,
+    global_load_sum: f64,
+    w_prime: f64,
+    dcycle_only: bool,
```

Comment:
I wonder whether it'd be easier to use if the same struct contained the unadjusted and adjusted values or does it get unwieldy with prefixes? But, right now, if I want to compare the values before and after adjustments, I'd be comparing `load_aggregator.X` vs. `load_balancer.X` which isn't very intuitive. Wouldn't it be clearer if it's more explicit? e.g. maybe sth like `load_aggregator.in.X` vs `load_aggregator.eff.X`.